### PR TITLE
Improve CUDA extension building error/warning messages

### LINKED ISSUE #55267: Creating a cudatoolkit-dev to enable CUDA extension building
## 🚀 Feature

Requesting in addition to releasing `cudatoolkit` to also release `cudatoolkit-dev`.

## Motivation

It's so convenient that `pytorch` comes with the `cudatoolkit`.  Thank you to whoever had this bright idea in the first place!

Unfortunately when it comes to pytorch CUDA extensions one has to install a matching cuda system-wide, which is not always straightforward and there are all kinds of build problems around it. Very often JIT builds fail despite having the right cuda installed system-wide and there isn't always even the log why - often the build just hangs w/o any traceback.

But the most common problem is:
```
    Exception: Installed CUDA version 11.0 does not match the version torch was compiled with 10.2, 
unable to compile cuda/cpp extensions without a matching cuda version.
```

Would it be too difficult to also release `cudatoolkit-dev` which would enable building cuda extensions like `fairscale` and `deepspeed` w/o needing a matching system-wide cuda?

This is especially an issue for those of us that do a lot of testing and integration work and to be able to have various cuda toolkit builds contained to different virtual environments like conda would make such testing so much easier.

Thank you!

### PR DESCRIPTION
**Pull request by andfoy**
See #55267

### CODE REVIEW

**Comment by stas00** [comment]
I don't think this PR should auto-close https://github.com/pytorch/pytorch/issues/55267 - there are multiple issues being discussed there.

If we need to split them up into separate actionable Issues, that might be an easier approach to work with tracking-wise.

**Comment by andfoy** [comment]
@stas00 , this is a PR in progress that aims to fix all the posed cases discussed on the issue, ergo it is still in Draft. Right now we are discussing with @rgommers which scenario cases should raise errors and which warnings. If you could provide further information about the expected behaviour and cases, it would be appreciated.

File: torch/utils/cpp_extension.py [hunk 1]
```diff
@@ -740,6 +753,16 @@ def _check_abi(self):
...
             raise UserWarning(msg)
 
+    def _check_cuda_version(self):
+        if CUDA_HOME:
+            nvcc = os.path.join(CUDA_HOME, 'bin', 'nvcc')
+            cuda_version_str = subprocess.check_output([nvcc, '--version']).strip().decode()
+            cuda_version = re.search(r'release (\d+[.]\d+)', cuda_version_str)
+            if cuda_version is not None:
+                cuda_version = cuda_version.group(1)
+                if cuda_version != torch.version.cuda:
```
Comment (stas00) [inline torch/utils/cpp_extension.py:763]: Shouldn't this assert if:
1. cuda is None - you can't build CUDA extensions w/o system-wide cuda
2. major version mismatches?

Comment (andfoy) [reply torch/utils/cpp_extension.py:763]: Thanks for the quick review @stas00! 

1. I should verify if there is at least one CUDA extension to compile before checking if the warning should appear or not, otherwise it won't make much sense when compiling a C++ only extension.
2. True, this check should be more detailed

**Comment by stas00** [comment]
Understood! Thank you for doing that, @andfoy!

> If you could provide further information about the expected behaviour and cases, it would be appreciated.

as all cases at some point had a workaround, I don't have any outstanding issues that couldn't be worked around - albeit the main problem is the same - difficulty of use for non-advanced users. 

So we may need to raise each one and then will try to recall what the problem was. I'm sure for each type a handful of Issues could be found in pytorch and the projects that use it.

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -740,6 +778,19 @@ def _check_abi(self):
...
+    def _check_cuda_version(self):
+        if CUDA_HOME:
+            nvcc = os.path.join(CUDA_HOME, 'bin', 'nvcc')
+            cuda_version_str = subprocess.check_output([nvcc, '--version']).strip().decode()
+            cuda_version = re.search(r'release (\d+[.]\d+)', cuda_version_str)
+            if cuda_version is not None:
+                cuda_version = parse_version(cuda_version.group(1))
+                torch_cuda_version = parse_version(torch.version.cuda)
+                if cuda_version.major != torch_cuda_version.major:
+                    warnings.warn(CUDA_MISMATCH_MESSAGE.format(cuda_version, torch.version.cuda))
```
Comment (stas00) [inline torch/utils/cpp_extension.py:790]: why not assert here? we know it won't work.

File: torch/utils/cpp_extension.py [hunk 3]
```diff
@@ -740,6 +778,19 @@ def _check_abi(self):
...
+        else:
+            warnings.warn(CUDA_NOT_FOUND_MESSAGE)
```
Comment (stas00) [inline torch/utils/cpp_extension.py:792]: same here? why issue a warning when it can tell the user what to do right away to fix the problem

**Comment by rgommers** [comment]
> I don't think this PR should auto-close #55267 - there are multiple issues being discussed there.

I agree. It will be good for this PR to only tackle the CUDA mismatch issue, and have separate PRs for the cache overwriting, `g++` and env vars issues - one PR per topic.

File: torch/utils/cpp_extension.py [hunk 4]
```diff
@@ -740,6 +763,21 @@ def _check_abi(self):
                    'Please set `DISTUTILS_USE_SDK=1` and try again.')
             raise UserWarning(msg)
 
+    def _check_cuda_version(self):
+        if CUDA_HOME:
```
Comment (rgommers) [inline torch/utils/cpp_extension.py:767]: I don't think this is a robust check. For example, on my current Arch Linux setup:
```bash
$ echo $CUDA_HOME

$ which nvcc
/opt/cuda/bin/nvcc
rgommers@machine:~/code/numpy.org (master)$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
...
Build cuda_11.3.r11.3/compiler.29745058_0
```
There's a more elaborate function `_find_cuda_home` around line 50 in this file.

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -740,6 +763,21 @@ def _check_abi(self):
...
+            nvcc = os.path.join(CUDA_HOME, 'bin', 'nvcc')
+            cuda_version_str = subprocess.check_output([nvcc, '--version']).strip().decode()
+            cuda_version = re.search(r'release (\d+[.]\d+)', cuda_version_str)
+            if cuda_version is not None:
+                cuda_str_version = cuda_version.group(1)
+                cuda_version = parse_version(cuda_str_version)
+                torch_cuda_version = parse_version(torch.version.cuda)
+                if cuda_version.major != torch_cuda_version.major:
```
Comment (rgommers) [inline torch/utils/cpp_extension.py:775]: This is the first part of https://github.com/pytorch/pytorch/issues/55267#issuecomment-860859575. Shouldn't it be followed by
```python
elif cuda_version.minor != torch_cuda_version.minor:
    warnings.warn(....)
```
?

File: torch/utils/cpp_extension.py (hunk 4 above)
Comment (andfoy) [reply torch/utils/cpp_extension.py:767]: The problem is that one checks the existence of CUDA in the system by locating nvcc, not checking its version. This check looks for the version listed in line 4, for example:

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Mon_Nov_30_19:08:53_PST_2020
Cuda compilation tools, release 11.2, V11.2.67
Build cuda_11.2.r11.2/compiler.29373293_0
```

The regex will parse the 11.2 string in `Cuda compilation tools, release 11.2, V11.2.67`

Comment (rgommers) [reply torch/utils/cpp_extension.py:767]: Sure, but my main point is you are raising an exception unless `CUDA_HOME` is set. I think for the Ubuntu et al. default location of CUDA, it's often found by default and the user doesn't have to set the env var. So that'll break, right?

Comment (andfoy) [reply torch/utils/cpp_extension.py:779]: The `CUDA_HOME` variable is set by calling `_find_cuda_home`: https://github.com/pytorch/pytorch/blob/0c916c8a4e94c50ae65c2acdfed348cb94bd26de/torch/utils/cpp_extension.py#L163

File: torch/utils/cpp_extension.py [hunk 6]
```diff
@@ -157,6 +157,18 @@ def _join_rocm_home(*paths) -> str:
...
+                               !! WARNING !!
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+The detected CUDA version ({0}) mismatches the version that was used to compile
+PyTorch ({1}). Please make sure to use the same CUDA versions, otherwise some
+errors may appear during compilation/runtime.
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+                              !! WARNING !!
+'''
```
Comment (stas00) [inline torch/utils/cpp_extension.py:171]:
```suggestion
CUDA_MISMATCH_WARN = "The detected CUDA version ({0}) mismatches the version that was used to compile
PyTorch ({1}). Most likely this shouldn't be a problem"
```

As discussed in the original feedback thread this is not user-friendly - printing the huge full screen WARNING is super-scary for users and they will be seeking help.

Every time there is any problem with CUDA extension users tend to quote these WARNINGS, since they look like errors. 

If we agreed that most of the time minor version mismatch should be not a problem, let's make it so. Hence the  proposed low-profile warning.

Comment (stas00) [reply torch/utils/cpp_extension.py:171]: except I don't know if it was used elsewhere, so perhaps this one should be called explicitly `CUDA_MISMATCH_MINOR_VERSION_WARN`

File: torch/utils/cpp_extension.py [hunk 7]
```diff
@@ -153,6 +153,15 @@ def _join_rocm_home(*paths) -> str:
 
                               !! WARNING !!
 '''
+CUDA_MISMATCH_MESSAGE = '''
+The detected CUDA version ({0}) mismatches the version that was used to compile
+PyTorch ({1}). Please make sure to use the same CUDA versions.
+'''
+CUDA_MISMATCH_WARN = "The detected CUDA version ({0}) mismatches the minor version that was used to compile PyTorch ({1}). Most likely this shouldn't be a problem."
```
Comment (stas00) [inline torch/utils/cpp_extension.py:160]:
```suggestion
CUDA_MISMATCH_WARN = "The detected CUDA version ({0}) has a minor version mismatch with the version that was used to compile PyTorch ({1}). Most likely this shouldn't be a problem."
```
the original version was a bit hard to parse, here is a better wording.

File: torch/utils/cpp_extension.py [hunk 8]
```diff
@@ -153,6 +153,15 @@ def _join_rocm_home(*paths) -> str:
 
                               !! WARNING !!
 '''
+CUDA_MISMATCH_MESSAGE = '''
+The detected CUDA version ({0}) mismatches the version that was used to compile
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:157]: nit: consider using named interpolates rather than positional. Also, I know the existing code has the warning out of line, but since we've got f-strings now, consider putting the strings directly inline in the site where they're used; I typically find it easier to understand the code when everything is in one place. (Assuming everything else in the PR is fine I plan to merge this as is.)

**Review by ezyang — APPROVED** [review]

File: torch/utils/cpp_extension.py [hunk 9]
```diff
@@ -740,6 +764,24 @@ def _check_abi(self):
...
             raise UserWarning(msg)
 
+    def _check_cuda_version(self):
+        if CUDA_HOME:
+            nvcc = os.path.join(CUDA_HOME, 'bin', 'nvcc')
+            cuda_version_str = subprocess.check_output([nvcc, '--version']).strip().decode()
+            cuda_version = re.search(r'release (\d+[.]\d+)', cuda_version_str)
+            if cuda_version is not None:
+                cuda_str_version = cuda_version.group(1)
+                cuda_ver = parse_version(cuda_str_version)
```
Comment (walterddr) [inline torch/utils/cpp_extension.py:774]: apparently this returns a type of `pkg_resources.extern.packaging.version.Version` instead of `packaging.version.Version` and seems on some version it doesn't support `.major/minor` operation. changing it back to using `packaging.version.parse`
